src/history/hs.c

```diff
@@ -1239,3 +1239,123 @@ __hs_delete_key_from_pos(
+         */
+        cursor->set_key(cursor, btree_id_given_uri, hs_key, 0, 0, 0, 0);
+        ret = cursor->search_near(cursor, &exact);
+        if (ret == 0 && exact < 0)
+            ret = cursor->next(cursor);
+    } else
+        ret = cursor->next(cursor);
+
+    /* We have the history store cursor positioned at the first record that we want to verify. */
+    for (; ret == 0; ret = cursor->next(cursor)) {
+        WT_ERR(cursor->get_key(cursor, &btree_id, hs_key, &hs_start_ts, &hs_counter));
+
+        /* When limiting our verification to a uri, bail out if the btree-id doesn't match. */
+        if (uri != NULL && btree_id != btree_id_given_uri)
+            break;
+
+        /*
+         *  Keep track of the previous comparison. The history store is stored in order, so we can
+         *  avoid redundant comparisons. Previous btree ID isn't set, until data cursor is open.
+         */
+        if (data_cursor == NULL || (prev_btree_id != btree_id)) {
+            /*
+             * Check whether this btree-id exists in the metadata. We do that by finding what uri
+             * this btree belongs to. Using this URI, verify the history store key with the data
+             * store.
+             */
+            if (data_cursor != NULL) {
+                WT_ERR(data_cursor->close(data_cursor));
+                /* Setting data_cursor to null, to avoid double free */
+                data_cursor = NULL;
+            }
+            /*
+             * Using the btree-id find the metadata entry and extract the URI for this btree. Don't
+             * forget to free the copy of the URI returned.
+             *
+             * Re-purpose the previous-key buffer on error, safe because we're about to error out.
+             */
+            __wt_free(session, uri_itr);
+            if ((ret = __wt_metadata_btree_id_to_uri(session, btree_id, &uri_itr)) != 0)
+                WT_ERR_MSG(session, ret,
+                  "Unable to find btree-id %" PRIu32
+                  " in the metadata file for the associated history store key %s",
+                  btree_id,
+                  __wt_buf_set_printable(session, hs_key->data, hs_key->size, prev_hs_key));
+
+            WT_ERR(__wt_open_cursor(session, uri_itr, NULL, NULL, &data_cursor));
+            F_SET(data_cursor, WT_CURSOR_RAW_OK);
+        } else {
+            WT_ERR(__wt_compare(session, NULL, hs_key, prev_hs_key, &cmp));
+            if (cmp == 0)
+                continue;
+        }
+        WT_ERR(__wt_buf_set(session, prev_hs_key, hs_key->data, hs_key->size));
+        prev_btree_id = btree_id;
+
+        /* Re-purpose the previous-key buffer on error, safe because we're about to error out. */
+        data_cursor->set_key(data_cursor, hs_key);
+        if ((ret = data_cursor->search(data_cursor)) == WT_NOTFOUND)
+            WT_ERR_MSG(session, ret,
+              "In %s, the associated history store key %s not found in the data store", uri_itr,
```

Comment:
Maybe "was not found" instead of just not found.

Reply:
Pushed in 66ccd6594.